### PR TITLE
Adds support for input[type=file] for enctype=multipart/form-data

### PR DESCRIPTION
**Pull request by joyheron**
My hypothesis here is that when the user needs to use an
input[type=file] they will also need to use a form which is has the
enctype `multipart/form-data` (and also makes the assumption that
the form method is not 'GET', because a form upload should not be mapped
to a GET resource). This modifies the `serializeForm` method to add two
different accumulators: one for collecting the values of the input
fields as string values and then encoding them as a stringified
application/x-www-form-urlencoded URL, and one which collects them as a
`FormData` object which can then be directly set as the body of the http
request.

The new serializeForm method is still backwards compatible because it
uses a StringParams() accumulator by default.

Note: I haven't tested this implementation very carefully (yet), so it
should be carefully reviewed and tested before merging.

### CODE REVIEW

File: src/util.js [hunk 1]
```diff
@@ -58,7 +60,8 @@ export function serializeForm(form) {
 		case "input":
 			switch(node.type) {
 			case "file":
-				throw new Error("`input[type=file]` fields are unsupported");
+				value = node.files[0];
```
Comment (FND) [inline src/util.js:63]:
```suggestion
				// NB: due to `enctype` differentiation, files and
				//     `StringParams` should be mutually exclusive
				value = node.files[0];
```

also, what about `<input type="file" multiple>`; there'd be more then one file there, no?

File: src/util.js [hunk 2]
```diff
@@ -87,12 +90,45 @@ export function serializeForm(form) {
...
-	}, []).join("&");
+	}, params).serialize();
+}
+
+class FormDataParams {
+	constructor() {
+		this.formData = new FormData();
+	}
+
+	setParameter(name, value) {
```
Comment (FND) [inline src/util.js:105]:
```suggestion
	add(name, value) {
```

(same for `StringParams`, of course)

File: src/util.js [hunk 3]
```diff
@@ -87,12 +90,45 @@ export function serializeForm(form) {
...
+		this.formData.append(name, value);
+	}
+
+	serialize() {
```
Comment (FND) [inline src/util.js:109]: "serialize" seems misleading when returning a `FormData` instance? I'm not sure what might be a good term though; "collate", _maybe_?

(I noticed because I wanted to suggest [`s/serialize/toString/`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/toString#Description) for `StringParams` below)

File: src/util.js [hunk 4]
```diff
@@ -87,12 +90,45 @@ export function serializeForm(form) {
...
+	}
+}
+
+class StringParams {
+	constructor() {
+		this.params = [];
+	}
+
+	setParameter(name, value) {
+		if(typeof (value) !== "string") {
```
Comment (FND) [inline src/util.js:120]: when might this occur?

File: src/util.js [hunk 5]
```diff
@@ -87,12 +90,45 @@ export function serializeForm(form) {
...
+			throw new Error("`input[type=file]` fields are only supported for " +
+				"forms with [enctype=multipart/form-data]");
```
Comment (FND) [inline src/util.js:122]: FNDism:

```suggestion
			throw new Error("`input[type=file]` fields are only supported " +
					"for forms with [enctype=multipart/form-data]");
```

File: src/util.js [hunk 6]
```diff
@@ -6,18 +6,20 @@ export function submitForm(form, { headers, cors, strict } = {}) {
 	let method = form.getAttribute("method");
 	method = method ? method.toUpperCase() : "GET";
 	let uri = form.getAttribute("action");
-	let payload = serializeForm(form);
+	let body;
 
 	if(method === "GET") {
```
Comment (FND) [inline src/util.js:11]:
```suggestion

	let body;
	if(method === "GET") {
```

(though personally, I'd still prefer the `var` hoisting hack)

**Review by FND — COMMENTED** [review]
This seems pretty cool, thank you very much!

I haven't had a chance to play with this yet, so just a few preliminary comments.

File: src/util.js (hunk 6 above)
Comment (joyheron) [reply src/util.js:11]: But does the `var` hoisting hack work if you have two separate `var body = serializeForm()` statements?

File: src/util.js (hunk 4 above)
Comment (joyheron) [reply src/util.js:120]: Probably a leaky abstraction 🙈 

This occurs if the user calls `submitForm` on a `form` element which does not have the correct `enctype` set. Essentially, when using the `StringParams` accumulator you need to make sure that all of your input parameters can actually be persisted as strings.

File: src/util.js (hunk 6 above)
Comment (FND) [reply src/util.js:11]: Yes: It essentially translates to adding `let body;` as the function's very first line.

You'd only need `var` once though.

File: src/util.js (hunk 3 above)
Comment (joyheron) [reply src/util.js:109]: Naming is hard. Originally I had named this `toFormBody()`, but I changed that because for `GET` forms the result of `serializeForm` goes in the URI, not in the body. 

I'm not against "collate", but it is an obscure enough word that I have to look up its definition every time I see it.

File: src/util.js (hunk 2 above)
Comment (joyheron) [reply src/util.js:105]: Yes, that seems like a good idea.

File: src/util.js (hunk 1 above)
Comment (joyheron) [reply src/util.js:63]: I of course did not test that 🙈 

I'm wondering if we could just use `value = node.files` here, since we iterate over `value` in a later step

File: src/util.js (hunk 6 above)
Comment (joyheron) [reply src/util.js:11]: And it would probably break if you add `var` twice, correct?

Comment (FND) [reply src/util.js:11]: Depends: JavaScript would essentially ignore the second `var` - but ESLint would complain. 🙂

Comment (joyheron) [reply src/util.js:11]: It's your library. I could live with either.

Comment (FND) [reply src/util.js:11]: Ma'am, I believe there's a misunderstanding here: Clearly this is now your library.

File: src/util.js [hunk 7]
```diff
@@ -6,18 +6,20 @@ export function submitForm(form, { headers, cors, strict } = {}) {
...
 		if(uri.indexOf("?") !== -1) {
 			throw new Error("query strings are invalid within `GET` forms' " +
 					"`action` URI; please use hidden fields instead");
 		}
-		uri = [uri, payload].join("?");
+		uri = [uri, serializeForm(form, new StringParams())].join("?");
+	} else if(form.getAttribute("enctype") === "multipart/form-data") {
```
Comment (FND) [inline src/util.js:17]: While generally speaking, this heuristic should be fine, there's also the option to [override the form's attribute via `formenctype`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#Attributes_for_form_submission) - not sure whether that's something we could detect or have to account for; seems like a very rare edge case, at least. (See README for [related caveats](https://github.com/FND/hijax-form#caveats).)

File: src/util.js (hunk 1 above)
Comment (FND) [reply src/util.js:63]: Based on a cursory check, it doesn't seem to work that way:

```html
    <form>
        <input type="file" name="attachments" multiple>
    </form>
    <script>
document.querySelector("input").addEventListener("change", ev => {
    let field = ev.target;
    console.log(field.files);

    let params = new FormData();
    params.append(field.name, field.files);
    console.log(params, [...params.keys()], [...params.values()]);
});
    </script>
```

Assigning `field.files` instead of `field.files[0]` there appears to result in stringification: `"[object FileList]"` 🤷

But I guess iterating over `field.files` and invoking `#append` individually should work?

Comment (joyheron) [reply src/util.js:63]: I would hope that iterating and invoking #append individually would work.

that is what happens here anyway: https://github.com/FND/hijax-form/blob/master/src/util.js#L89, so I would hope that we wouldn't have to change too much in the implementation

File: src/util.js (hunk 3 above)
Comment (FND) [reply src/util.js:109]: As discussed elsewhere, perhaps we should just use a getter:

```javascript
    constructor() {
        this._data = [];
    }

    get data() {
        return this._data.join("&");
    }
```
```javascript
    constructor() {
        this.data = new FormData();
    }
```

File: src/util.js (hunk 4 above)
Comment (FND) [reply src/util.js:120]: uhm... this makes me realize that perhaps `serializeForm` isn't even warranted when using `FormData`: doesn't `FormData` itself already provide that very functionality? i.e. [`new FormData(formNode)` should suffice](https://developer.mozilla.org/en-US/docs/Web/API/FormData/FormData)

Comment (joyheron) [reply src/util.js:120]: It's possible. that would be nice.

Comment (FND) [reply src/util.js:120]: I'm pretty sure that's how it was designed - to be submitted like this:

```javascript
fetch(uri, {
    method: "post",
    body: new FormData(formNode),
    …
})
```

which of course means that `FormData` only makes sense if native `fetch` is available (though that should always be the case IIRC)

either way, we should be explicit about what the progressive-enhancement story here is: at what point do we abort and just let native (i.e. non-AJAX) form submissions kick in? is there anything in between, like `fetch`/`FormData` being unavailable (think IE 11) - in which case stringification would still be an option, files excluded

in other words: for anything other than IE 11, most of the code in this module is unnecessary

Comment (joyheron) [reply src/util.js:120]: https://caniuse.com/?search=formdata

FormData is supported on IE11 (at least it says that it is)

Comment (joyheron) [reply src/util.js:120]: I've just tested the implementation in IE11 that I am currently using (which uses `FormData` when the `enctype` is correct) and in that case, the upload with ajax works, so I would say that we can probably safely use `FormData`.

File: src/util.js [hunk 8]
```diff
@@ -6,18 +6,20 @@ export function submitForm(form, { headers, cors, strict } = {}) {
...
+		body = serializeForm(form, new FormDataParams());
```
Comment (joyheron) [inline src/util.js:18]: I need to test this out, but theoretically we should be able to use `body = new FormData(form)` here. Then it is not necessary to call the `serializeForm` function
